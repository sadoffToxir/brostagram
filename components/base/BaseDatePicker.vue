<template>
  <div :data-test="componentName">
    <v-text-field
      :model-value="date"
      variant="underlined"
      prepend-inner-icon="mdi-calendar-range"
      readonly
      :label="label"
      :hide-details="hideDetails"
      :disabled="disabled"
      :error="error"
      :data-test="componentName + '-input'"
      :error-messages="errorMessages"
      class="baseDatePicker-trigger"
      @click="(baseDatePicker as any).openMenu()"
    />
    <VueDatePicker
      ref="baseDatePicker"
      uid="baseDatePicker"
      :min-date="min"
      :model-value="date"
      :enable-time-picker="false"
      :month-change-on-scroll="false"
      :max-date="max"
      :data-test="componentName + '-datePicker'"
      auto-apply
      :teleport="true"
      @update:model-value="handleChange"
    >
      <template #trigger />
    </VueDatePicker>
  </div>
</template>

<script setup lang="ts">
interface Props {
  date: string
  min?: string | Date
  label?: string
  max?: string | Date
  disabled?: boolean
  hideDetails?: boolean | 'auto'
  error?: boolean
  errorMessages?: string | Array<string>
}

withDefaults(defineProps<Props>(), {
  label: 'Date',
  min: '',
  max: '',
  hideDetails: 'auto',
  error: false,
  errorMessages: '',
  disabled: false
})

interface Emits {
  (e: 'update:date', value: string): void
}

const emits = defineEmits<Emits>()

const baseDatePicker = ref(null)
const componentName = getCurrentInstance()?.type.__name

const handleChange = (value: Date) => {
  emits('update:date', formatToIsoDate(value))
}
</script>

<style lang="scss">
.baseDatePicker-trigger {
  .v-field__prepend-inner {
    align-items: center;
  }
}
</style>
