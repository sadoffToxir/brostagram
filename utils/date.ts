type CustomDate = string | number | Date

export const formatToIsoDate = (date: CustomDate) => new Date(date).toISOString().split('T')[0]
export const formatToIsoDateAndTime = (date: CustomDate) => new Date(date).toISOString().split('T')[0] + ' at ' + new Date(date).toISOString().split('T')[1].substring(0, 5)
