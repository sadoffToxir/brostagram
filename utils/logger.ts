import { WritableComputedRef, Ref } from 'vue'

export default function logger (consoleText: string, observed: WritableComputedRef<any> | Ref<any>) {
  const config = useRuntimeConfig()

  if (config.public.environment !== 'production') {
    watch(observed, useDebounce(() => {
    // eslint-disable-next-line no-console
      console.log(`${consoleText}:`, useCloneDeep(observed.value))
    }, 500), { deep: true })
  }
}
