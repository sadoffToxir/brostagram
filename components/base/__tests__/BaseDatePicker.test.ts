import VueDatePicker from '@vuepic/vue-datepicker'
import { describe, it, expect } from 'vitest'
import { fireEvent } from '@testing-library/vue'
import BaseDatePicker from '../BaseDatePicker.vue'
import { renderWithVuetify, screen } from '@/tests/vueTestingLibrary/utils'
import { formatToIsoDate } from '@/utils/date'

const componentName = BaseDatePicker.__name as string

describe(componentName, () => {
  const globalOptions = {
    global: {
      components: {
        VueDatePicker
      }
    }
  }

  const setup = (options?: Record<string, object>) => renderWithVuetify(BaseDatePicker, {
    global: {
      ...globalOptions.global,
      ...options?.global
    },
    ...options
  })

  it('renders with props', () => {
    const props = {
      label: 'label',
      date: formatToIsoDate(new Date())
    }

    setup({ props })

    const element = screen.getByTestId(componentName)
    const label = screen.getByLabelText(props.label)
    const date = screen.getByDisplayValue(props.date)

    expect(element).toBeVisible()
    expect(date).toBeVisible()
    expect(label).toBeVisible()
  })

  it('triggers calendar modal on input click', async () => {
    const props = {
      date: formatToIsoDate(new Date())
    }

    setup({ props })

    const input = screen.getByTestId(`${componentName}-input`).querySelector('input') as Element
    await fireEvent.click(input)
    const dialog = document.body.querySelector('#dp-menu-baseDatePicker')

    expect(dialog).toBeVisible()
  })

  it('can select date', async () => {
    const date = formatToIsoDate('2023-08-16')
    const activeDate = new Date(date).getDate()
    const props = {
      date
    }

    const { emitted } = setup({ props })

    // check if dialog opens and date from input is active date in the dialog
    const input = screen.getByTestId(`${componentName}-input`).querySelector('input') as Element
    await fireEvent.click(input)
    const dialog = document.body.querySelector('#dp-menu-baseDatePicker')
    expect(dialog).toBeInTheDocument()
    const active = dialog!.querySelector('.dp__active_date')
    expect(active).toHaveTextContent(String(activeDate))

    // select day before active date and check if dialog closes + value is updated
    const previous = screen.getByText(String(activeDate - 1))
    await fireEvent.click(previous)
    expect(document.body.querySelector('#dp-menu-baseDatePicker')).not.toBeInTheDocument()
    expect(emitted('update:date')).toMatchSnapshot()
  })
})
