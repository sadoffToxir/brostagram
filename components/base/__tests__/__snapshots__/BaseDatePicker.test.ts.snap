// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseDatePicker > can select date 1`] = `
[
  [
    "2023-08-15",
  ],
]
`;
